### PR TITLE
libvirt: use pure go implementation in static mode

### PR DESCRIPTION
**Pull request by lebauce**

### CODE REVIEW

**Comment by lebauce** [comment]
run skydive-functional-tests-backend-orientdb

File: Makefile [hunk 1]
```diff
@@ -89,7 +89,7 @@ BOOTSTRAP_ARGS?=
 BUILD_TAGS?=$(TAGS)
 WITH_LXD?=true
 WITH_OPENCONTRAIL?=true
-WITH_LIBVIRT?=true
+WITH_LIBVIRT_GO?=true
```
Comment (hunchback) [inline Makefile]: why does it matter how its' implemented? rather leave the flag name `WITH_LIBVIRT`

File: topology/probes/libvirt/libvirt.go [hunk 2]
```diff
@@ -39,16 +39,27 @@ import (
 type Probe struct {
 	sync.Mutex
 	graph        *graph.Graph          // the graph
-	conn         *libvirtgo.Connect    // libvirt conection
+	conn         monitor               // libvirt conection
```
Comment (hunchback) [inline topology/probes/libvirt/libvirt.go]: s/libvirt conection/libvirt connection/

File: topology/probes/libvirt/libvirtgo.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+ *
+ */
+
+package libvirt
+
+import (
+	"context"
+	"fmt"
+
+	libvirtgo "github.com/libvirt/libvirt-go"
```
Comment (hunchback) [inline topology/probes/libvirt/libvirtgo.go]: not consistent binding either `libvirtgo` or `libvirt` used consistently throughout the project

File: topology/probes/libvirt/libvirtgo.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+)
+
+type libvirtgoDomain struct {
+	libvirtgo.Domain
+}
+
+func (d libvirtgoDomain) GetName() (string, error) {
+	return d.Domain.GetName()
+}
+func (d libvirtgoDomain) GetXML() ([]byte, error) {
```
Comment (hunchback) [inline topology/probes/libvirt/libvirtgo.go]: Missing an empty line between functions

File: topology/probes/libvirt/golibvirt.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+	return allDomains, nil
+}
+
+func (m *golibvirtMonitor) Stop() {
+	m.libvirt.Disconnect()
+}
+
+func newMonitor(ctx context.Context, probe *Probe) (*golibvirtMonitor, error) {
+	if probe.uri == "" {
+		probe.uri = "unix:///var/run/libvirt/libvirt-sock"
```
Comment (hunchback) [inline topology/probes/libvirt/golibvirt.go]: You probably want to make this uri configurable as location may vary on different systems: https://www.redhat.com/archives/libvirt-users/2012-October/msg00137.html

File: topology/probes/libvirt/golibvirt.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+	conn, err := net.DialTimeout(split[0], split[1], 2*time.Second)
+	if err != nil {
+		return nil, err
+	}
+
+	libvirt := golibvirt.New(conn)
+	monitor := &golibvirtMonitor{libvirt: libvirt}
+
+	if err := libvirt.Connect(); err != nil {
+		return nil, fmt.Errorf("failed to connect: %v", err)
```
Comment (hunchback) [inline topology/probes/libvirt/golibvirt.go]: s/%v/%s/

File: topology/probes/libvirt/libvirtgo.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+	}
+
+	if _, err := m.Close(); err != nil {
+		logging.GetLogger().Errorf("Problem during close: %s", err)
+	}
+}
+
+func newMonitor(ctx context.Context, probe *Probe) (*LibvirtgoMonitor, error) {
+	if probe.uri == "" {
+		probe.uri = "qemu:///system"
```
Comment (hunchback) [inline topology/probes/libvirt/libvirtgo.go]: You may consider making this configurable: https://libvirt.org/uri.html#URI_qemu

**Comment by hunchback** [comment]
@lebauce - regarding file naming - I wanted to propose the following:
- libvirt.go - entry point
- rpc.go - using the go native library which uses the RPC to communicate to libvirtd
- sock.go - the C based solution accessing the `/var/run/libvirt/libvirt-sock`

File: Makefile (hunk 1 above)
Comment (lebauce) [reply Makefile]: With this PR, we will be able to always enable the `libvirt` probe. WITH_LIBVIRT_GO will specify if we want to use the `libvirt-go` library (that includes a dynamic dependency on `libvirt.so`). When set to false, the pure Go implementation will be used. It has deployment implication so I think it makes sense to define a variable for it

Comment (lebauce) [reply Makefile]: For the name of the variable itself, reusing WITH_LIBVIRT may let the user think that setting it to false will disable the `libvirt` probe.

File: topology/probes/libvirt/libvirtgo.go (hunk 4 above)
Comment (lebauce) [reply topology/probes/libvirt/libvirtgo.go]: Fixed

File: topology/probes/libvirt/golibvirt.go (hunk 6 above)
Comment (lebauce) [reply topology/probes/libvirt/golibvirt.go]: Fixed

File: topology/probes/libvirt/libvirtgo.go (hunk 7 above)
Comment (lebauce) [reply topology/probes/libvirt/libvirtgo.go]: This is configurable as `uri` comes from the configuration file. If it's empty - and for the libvirt-go based implementation - we use fallback to `qemu:///system`

**Comment by lebauce** [comment]
@hunchback Regarding the file naming, I'm not a fan of rpc.go and sock.go as both implementations use sockets (and rpc under the hood). The difference is really about the Go library we use.

File: topology/probes/libvirt/golibvirt.go (hunk 5 above)
Comment (lebauce) [reply topology/probes/libvirt/golibvirt.go]: `uri` comes from the configuration file, so it's configurable

**Comment by lebauce** [comment]
run skydive-cdd-overview-tests

**Comment by lebauce** [comment]
run skydive-functional-hw-tests

**Comment by lebauce** [comment]
run skydive-functional-tests-backend-orientdb
run skydive-functional-hw-tests

**Comment by lebauce** [comment]
run skydive-functional-hw-tests

**Comment by lebauce** [comment]
run skydive-functional-hw-tests

**Comment by lebauce** [comment]
run skydive-functional-hw-tests

**Pull request merged**